### PR TITLE
Mfp read plain

### PR DESCRIPTION
**Pull request by merlokk**
`hf mfp` can `rdbl` and `rdsc` in plain mode
plus:
- aescmac128
- tests for aescmac

TODO:
- move from **polarssl** to **mbedtls**

_I dont know the MAC calc procedure. so i leave `wrbl` command to time when it will be possible._

### CODE REVIEW

**Comment by merlokk** [comment]
merge plz)

File: client/mifare4.c [hunk 1]
```diff
@@ -16,6 +16,18 @@
 #include "ui.h"
 #include "polarssl/libpcrypto.h"
 
+int CalulateMAC(mf4Session *session, uint8_t *data, int datalen, uint8_t *mac, bool verbose) {
```
Comment (pwpiwi) [inline client/mifare4.c]: Typo: CalculateMAC()

File: client/mifare4.c [hunk 2]
```diff
@@ -16,6 +16,18 @@
...
+	if (!session || !session->Authenticated || !mac || !data || !datalen)
+		return 1;
+	
+	memset(mac, 0x00, 8);
+	
+	if (verbose)
+		PrintAndLog("MAC data[%d]: %s", datalen, sprint_hex(data, datalen));
+
```
Comment (pwpiwi) [inline client/mifare4.c]: see http://www.proxmark.org/forum/viewtopic.php?id=5348 for the input to CMAC calculation.

Comment (merlokk) [reply client/mifare4.c]: cool)

File: client/cmdhfmfp.c [hunk 3]
```diff
@@ -381,19 +415,302 @@ int CmdHFMFPAuth(const char *cmd) {
...
+	keyn[0] = uKeyNum >> 8;
+	keyn[1] = uKeyNum & 0xff;
+	if (verbose)
+		PrintAndLog("--sector[%d]:%02x key:%04x", mfNumBlocksPerSector(sectorNum), sectorNum, uKeyNum);
+	
+	mf4Session session;
+	int res = MifareAuth4(&session, keyn, key, true, true, verbose);
+	if (res) {
+		PrintAndLog("Authentication error: %d", res);
+		return res;
```
[the remark this answers is not in the document]
Comment (merlokk) [reply client/cmdhfmfp.c]: in `MifareAuth4` there is a Dropfield()

File: client/cmdhfmfp.c [hunk 4]
```diff
@@ -381,19 +415,302 @@ int CmdHFMFPAuth(const char *cmd) {
...
+	keyn[0] = uKeyNum >> 8;
+	keyn[1] = uKeyNum & 0xff;
+	if (verbose)
+		PrintAndLog("--block:%d sector[%d]:%02x key:%04x", blockNum & 0xff, mfNumBlocksPerSector(sectorNum), sectorNum, uKeyNum);
+	
+	mf4Session session;
+	int res = MifareAuth4(&session, keyn, key, true, true, verbose);
+	if (res) {
+		PrintAndLog("Authentication error: %d", res);
+		return res;
```
[the remark this answers is not in the document]
Comment (merlokk) [reply client/cmdhfmfp.c]: Dropfield done

**Comment by merlokk** [comment]
can be merged. p.s. look into mail plz

**Pull request merged**